### PR TITLE
Add readiness check for resources

### PR DESCRIPTION
**Pull request by michaelhtm**
Issue #16 

Description of changes:
This feature ensures users are allowed to define 
when a resource is ready or not.

For now this feature only allows to look within itself to 
see whether it's ready or not, so only its spec and status 
can be accessed using CEL fields.

An example with deployement-service RG will be included i
n this PR

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: api/v1alpha1/resource_group.go [hunk 1]
```diff
@@ -72,6 +72,8 @@ type Resource struct {
 	Name string `json:"name,omitempty"`
 	// +kubebuilder:validation:Required
 	Definition runtime.RawExtension `json:"definition,omitempty"`
+	// +kubebuilder:valisation:Optional
```
Comment (a-hilaly) [inline api/v1alpha1/resource_group.go]:
```suggestion
	// +kubebuilder:validation:Optional
```

File: internal/controller/instance/controller_reconcile.go [hunk 2]
```diff
@@ -212,6 +212,13 @@ func (igr *InstanceGraphReconciler) reconcileResource(ctx context.Context, resou
 		return resourceState.Err
 	}
 
+	log.V(1).Info("Checking if resource is Ready", "resource", resourceID)
+	if !igr.runtime.IsResourceReady(resourceMeta.ReadyOnExpressions, observed) {
```
Comment (a-hilaly) [inline internal/controller/instance/controller_reconcile.go]: The runtime has already access to these expressions, can we instead make it `	if !igr.runtime.IsResourceReady(resourceID) {`

File: internal/typesystem/parser/parser.go [hunk 3]
```diff
@@ -31,6 +31,24 @@ func ParseResource(resource map[string]interface{}, resourceSchema *spec.Schema)
 	return parseResource(resource, resourceSchema, "")
 }
 
+func ParseReadyOn(readyOn []string) ([]string, error) {
```
Comment (a-hilaly) [inline internal/typesystem/parser/parser.go]: Can we have some comments explaining what does this function do?

Comment (a-hilaly) [reply internal/typesystem/parser/parser.go]: Also needs unit tests :)

File: internal/resourcegroup/runtime.go [hunk 4]
```diff
@@ -104,6 +104,60 @@ func (rt *RuntimeResourceGroup) CanResolveResource(resource string) bool {
...
+			return false
+		}
+		out, err := celutil.ConvertCELtoGo(output)
+		if err != nil {
+			return false
+		}
+		// keep checking for a false
+		if !out.(bool) {
+			return false
+		}
```
Comment (a-hilaly) [inline internal/resourcegroup/runtime.go]: You already defined this in `celutil`

File: internal/resourcegroup/runtime.go [hunk 5]
```diff
@@ -104,6 +104,60 @@ func (rt *RuntimeResourceGroup) CanResolveResource(resource string) bool {
...
+	return true
+}
+
+func (rt *RuntimeResourceGroup) IsResourceReady(expressions []ReadyOnExpression, resource *unstructured.Unstructured) bool {
+
+	if len(expressions) == 0 {
+		return true
+	}
+
+	resourceNames := []string{"status", "spec"}
```
Comment (a-hilaly) [inline internal/resourcegroup/runtime.go]: are these resourceNames?

File: internal/resourcegroup/runtime.go [hunk 6]
```diff
@@ -104,6 +104,60 @@ func (rt *RuntimeResourceGroup) CanResolveResource(resource string) bool {
...
+		fmt.Println(e.Expression)
+		program, err := env.Program(ast)
+		if err != nil {
+			return false
+		}
+
+		output, _, err := program.Eval(context)
+		if err != nil {
+			return false
+		}
```
Comment (a-hilaly) [inline internal/resourcegroup/runtime.go]: I think we might want to Error here isn't of silently considering the resource not ready

File: internal/controller/instance/controller_reconcile.go [hunk 7]
```diff
@@ -212,6 +212,13 @@ func (igr *InstanceGraphReconciler) reconcileResource(ctx context.Context, resou
...
+		log.V(1).Info("Resource not ready", "resource", resourceID)
+		resourceState.State = "CREATING"
```
Comment (a-hilaly) [inline internal/controller/instance/controller_reconcile.go]: Should we introduce a state called `Pending`

File: examples/deployment-service/deployment-service.yaml [hunk 8]
```diff
@@ -14,13 +14,15 @@ spec:
       availableReplicas: ${deployment.status.availableReplicas}
   resources:
   - name: deployment
+    readyOn:
+      - ${spec.replicas == status.availableReplicas}
```
Comment (a-hilaly) [inline examples/deployment-service/deployment-service.yaml:18]: Can we also test this with EKS clusters? for example a Cluster need to be ready before we create a nodegroup. Or a vpc needs to be ready before we create subnets.

File: internal/controller/instance/controller_reconcile.go (hunk 7 above)
Comment (michaelhtm) [reply internal/controller/instance/controller_reconcile.go]: I thought `Pending` would be a state where the resource is waiting for other resources. In the case of readyOn, the resource is waiting to finish creating

Comment (a-hilaly) [reply internal/controller/instance/controller_reconcile.go]: You're right... then we need a better name here. `Progressing`, `Provisioning` ?

File: internal/resourcegroup/runtime.go (hunk 4 above)
Comment (michaelhtm) [reply internal/resourcegroup/runtime.go]: what i defined in celutil is only type checking, here I check if the output is false

File: internal/resourcegroup/runtime.go (hunk 5 above)
Comment (michaelhtm) [reply internal/resourcegroup/runtime.go]: no, wasn't sure what to name them
fields?

File: internal/controller/instance/controller_reconcile.go (hunk 7 above)
Comment (michaelhtm) [reply internal/controller/instance/controller_reconcile.go]: Why not creating?

Comment (michaelhtm) [reply internal/controller/instance/controller_reconcile.go]: `Reconciling`?

File: examples/ack-eks-cluster/eks-cluster.yaml [hunk 9]
```diff
@@ -56,6 +58,8 @@ spec:
           - destinationCIDRBlock: 0.0.0.0/0
             gatewayID: ${clusterInternetGateway.status.internetGatewayID}
   - name: clusterSubnetA
+    readyOn:
+    - ${status.state == "available"}
```
Comment (a-hilaly) [inline examples/ack-eks-cluster/eks-cluster.yaml:62]:
```suggestion
    - ${status.state == "ACTIVE"}
```

File: examples/ack-eks-cluster/eks-cluster.yaml [hunk 10]
```diff
@@ -19,6 +19,8 @@ spec:
   # resources
   resources:
   - name: clusterVPC
+    readyOn:
+    - ${status.state == "available"}
```
Comment (a-hilaly) [inline examples/ack-eks-cluster/eks-cluster.yaml:23]:
```suggestion
    - ${status.state == "ACTIVE"}
```

File: examples/ack-eks-cluster/eks-cluster.yaml [hunk 11]
```diff
@@ -169,6 +173,8 @@ spec:
               ]
           }
   - name: cluster
+    readyOn:
+    - ${status.status == "ACTIVE"}
```
Comment (a-hilaly) [inline examples/ack-eks-cluster/eks-cluster.yaml:177]: Were you able to test this one locally?

File: internal/controller/instance/controller_reconcile.go (hunk 7 above)
Comment (a-hilaly) [reply internal/controller/instance/controller_reconcile.go]: I think `Reconciling` might bring confusion, with the concepts of reconciling an RG/instance. How about `WaitingForReadiness`

File: internal/controller/instance/controller_reconcile.go [hunk 12]
```diff
@@ -316,6 +330,15 @@ func (igr *InstanceGraphReconciler) handleInstanceDeletion(ctx context.Context,
...
 		if resourceStates[resourceID].State == "PENDING_DELETION" {
+			// we need to:
+			// - get resource
+			// - if not found
+			//      continue
+			// - if found:
+			//      if metadata.deletionTimestamp == nil {
+			//          deleteResource{}
+			//      }
+			//      requeue
```
Comment (a-hilaly) [inline internal/controller/instance/controller_reconcile.go]: I think you forgot those :)

File: internal/resourcegroup/ready_on_expression.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+	fieldNames := []string{}
+
+	for k, _ := range schema.Properties {
+		if k != "apiVersion" && k != "kind" {
+			fieldNames = append(fieldNames, k)
+		}
+	}
+
+	slices.Sort(fieldNames)
```
Comment (a-hilaly) [inline internal/resourcegroup/ready_on_expression.go:46]: What could happen if they aren't sorted?

File: internal/resourcegroup/ready_on_expression.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+	Expression string
+}
+
+// This function goes through the schema of a resource
+// and retrieves the field names fo rthe readOnly feature
+//
+// Currently we support all the top level fields besides
+// apiVersion and kind. If we do see a case where they would
+// be necessary for readyOns, we can include them here.
+func getResourceTopLevelFieldNames(schema *spec.Schema) []string {
```
Comment (a-hilaly) [inline internal/resourcegroup/ready_on_expression.go:33]: 👍

File: internal/resourcegroup/runtime.go [hunk 15]
```diff
@@ -104,9 +104,64 @@ func (rt *RuntimeResourceGroup) CanResolveResource(resource string) bool {
 			return false
 		}
 	}
+
 	return true
 }
 
+func (rt *RuntimeResourceGroup) IsResourceReady(resourceID string) (bool, error) {
```
Comment (a-hilaly) [inline internal/resourcegroup/runtime.go:111]: Can you add a godoc for this function please?

File: internal/resourcegroup/runtime.go [hunk 16]
```diff
@@ -104,9 +104,64 @@ func (rt *RuntimeResourceGroup) CanResolveResource(resource string) bool {
...
+		ResourceNames: fieldNames,
+	})
+	// we should not expect errors here since we already compiled it
+	// in the dryRun
+	if err != nil {
+		return false, fmt.Errorf("failed creating new Environment: %w", err)
+	}
+	context := map[string]interface{}{}
+	for _, n := range fieldNames {
+		if obj, ok := resource.Object[n]; ok {
```
Comment (a-hilaly) [inline internal/resourcegroup/runtime.go:131]: how about naming this 
```suggestion
		if fieldValue, ok := resource.Object[n]; ok {
```

File: internal/typesystem/parser/conditions.go [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"fmt"
+	"strings"
+)
+
+// This function parses resource condition expressions.
+// These expressions need to be standalone expressions
+// so, this function also does some validation.
+// At the end we return the expressions with '${}' removed
```
Comment (a-hilaly) [inline internal/typesystem/parser/conditions.go:24]: I suggest to do a quick read on https://go.dev/blog/godoc

File: internal/typesystem/parser/conditions.go [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+//
+// To be honest I wouldn't necessarily call it parse, since
+// we are mostly just validating, without caring what's in 
+// the expression. Maybe we can rename it in the future 🤔
+func ParseConditionExpressions(conditions []string) ([]string, error) {
+
+	expressions := make([]string, 0, len(conditions))
+
+	for _, e := range conditions {
+		ok, err := isOneShotExpression(e)
```
Comment (a-hilaly) [inline internal/typesystem/parser/conditions.go:34]: should we renaming this function `isStandaloneExpression` ?

File: internal/typesystem/parser/conditions_test.go [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	"testing"
+)
+
+func TestParseReadyOn(t *testing.T) {
+	testCases := []struct {
+		name          string
+		expression    []string
+		expectedError string
+	}{
```
Comment (a-hilaly) [inline internal/typesystem/parser/conditions_test.go:25]: Can we also test empty strings, expressions with no quotes, and malformed expression

**Comment by a-hilaly** [comment]
@michaelhtm also golangci-lint has a few comments on this.

File: examples/ack-eks-cluster/eks-cluster.yaml (hunk 11 above)
Comment (michaelhtm) [reply examples/ack-eks-cluster/eks-cluster.yaml:177]: Yes, I was...nodegroup did wait for the cluster

File: examples/ack-eks-cluster/eks-cluster.yaml (hunk 10 above)
Comment (michaelhtm) [reply examples/ack-eks-cluster/eks-cluster.yaml:23]: the state for vpc and subnets is "available" 🤷‍♂️
the inconsistencies...
